### PR TITLE
Modify coordinate checking to support multi world and add Teleport() in Gateway

### PR DESCRIPTION
**Pull request by john0312**

### CODE REVIEW

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:194]: try **_occupying_** grid

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:195]: The use of **_whereby_** here is awkward... do you mean _**when**_?

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:196]: I think for now we should use a while loop to guard this function call instead of let it go.
```
while (! await this._occupyCoord(firstLocation.mapCoord, playerID));
```
Imagine that there is no spawn point that is not occupied. We should not let the player spawn since it might mess something later.

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:264]: Use `const` whenever possible. (maybe you forgot to run `eslint`?)

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:271]: "Overspeeds" is more clear than " is speeding".

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:268]: Do we really need this check?

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:280]: `eslint` will say you have bad spacing in this line.

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:363]: _**Nearby**_ is a word, so maybe it should be `_nearbyGridCheck`

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:284]: _**Nearby**_ is a word.

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:281]: Should we place the player right before the border instead of canceling their move?

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:290]: Do we really need this logging?

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:285]: `eslint`

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [reply services/gateway/gateway-service.mjs:363]: I don't see the need for this function. Isn't the functionality already been included in the speed check?

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:276]: I think this check should be included in the speed check.

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:291]: Why call `_teleportPlayerInternal` instead of `teleportPlayer`?

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:349]: I think this should be a utility function in `map.mjs` instead of a member function of `GatewayService`.

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:354]: Add jsdoc, please.
The available range of x coordinate is [0, mapSize.width), so the check should be
```
if (coord.x >= mapSize.width || ...)
```

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:357]: `>=` instead of `>`

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:381]: Where is the obstacle check? You should block players if they try to walk across a wall.

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:400]: Is `getRedis().setAsync()` going to return any value other than `OK` and `null`? I think you should find the documentation to ensure what value it is going to return.
```
const ret = await ...();
return (ret === 'OK');
```

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:407]: I think `if (playerID) {` is sufficient in this case.

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [inline services/gateway/gateway-service.mjs:411]: `if (getRet !== playerID) {`

**Review by ktpss95112 — CHANGES_REQUESTED** [review]
I think that these game-logic-related methods/functions should be inside `GameState` class. It is quite counterintuitive that we should look up `gateway-service.mjs` to find the implementation of moving a player. `gateway-service.mjs` should act as an interface only.
If I misunderstood something, please leave comments to let me know.

Also, add jsdoc, please. Run `eslint`, please.

File: services/gateway/gateway-service.mjs
Comment (john0312) [reply services/gateway/gateway-service.mjs:196]: Lisa's PR that's coming up soon will resolve this, so I would prefer leaving this here.

File: services/gateway/gateway-service.mjs
Comment (john0312) [reply services/gateway/gateway-service.mjs:268]: This is a guard so that we'll know when we screw up playerData.

File: services/gateway/gateway-service.mjs
Comment (john0312) [reply services/gateway/gateway-service.mjs:363]: This is needed because speed check checks only the time, this checks the distance.

File: services/gateway/gateway-service.mjs
Comment (john0312) [reply services/gateway/gateway-service.mjs:281]: I thought the two is the same?

File: services/gateway/gateway-service.mjs
Comment (john0312) [reply services/gateway/gateway-service.mjs:290]: It should happen very rarely in actual production. If it happens too frequently it'll point to something that we should pay attention to.

File: services/gateway/gateway-service.mjs
Comment (john0312) [reply services/gateway/gateway-service.mjs:276]: Will leave this to whyang when he moves everything that can be moved into the library.

File: services/gateway/gateway-service.mjs
Comment (john0312) [reply services/gateway/gateway-service.mjs:291]: Because we already have the socket and message.

File: services/gateway/gateway-service.mjs
Comment (john0312) [reply services/gateway/gateway-service.mjs:349]: I'll leave this to whyang.

File: services/gateway/gateway-service.mjs
Comment (john0312) [reply services/gateway/gateway-service.mjs:354]: I'll leave the task of adding jsdoc to whyang.

File: services/gateway/gateway-service.mjs
Comment (john0312) [reply services/gateway/gateway-service.mjs:381]: I'll leave this to whyang.

File: services/gateway/gateway-service.mjs
Comment (john0312) [reply services/gateway/gateway-service.mjs:400]: Generally it doesn't. The assertion is there to make sure it doesn't, and if it does due to any reasons, we'll know.

**Comment by john0312** [comment]
The purpose of this change is just to get whyang's previous PR to a state whereby it doesn't block the development. As discussed in the meeting, whyang will move the logics into GameState or another class in maplib. The task of adding jsdoc and fixing eslint will also be left to whyang.

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [reply services/gateway/gateway-service.mjs:282]: Not exactly. They are the same only if the player's speed is less than 1 cell per tick.
If the player can move at, say, 2 cells per tick, and the distance between him and the border is 1 cell, he would not able to move if we cancel his move. However, I would expect that he can move 1 cell (that is, right in front of the wall).

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [reply services/gateway/gateway-service.mjs]: I think we should use `assert` instead of `console.error` if our intention is to know when we screw up something.
```
assert('x' in lastCoord && 'y' in lastCoord && 'mapName' in lastCoord);
```

Comment (ktpss95112) [reply services/gateway/gateway-service.mjs]: I am wondering that `socket.playerData.mapCoord` is guaranteed to be a `class MapCoord` in `GatewayService.addSocket`. I mean, a crash of the server is reasonable if we really mess up something. We should not handle this error if this is not allowed to happen.

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [reply services/gateway/gateway-service.mjs]: The definition of speed is distance divided by time. The speed check should check whether a player **_moves how far_** within **_how much time_**.

File: services/gateway/gateway-service.mjs
Comment (ktpss95112) [reply services/gateway/gateway-service.mjs]: Isn't the documentation of Redis lists all possible return values?

**Comment by ktpss95112** [comment]
I understand that these changes are not complete features yet. However, I hope this merge is still robust and intuitive so that the code in the future will be in the correct direction.

**Comment by john0312** [comment]
For the task that this aims to unblock (Spawn Point), I don't think it'll affect the correct direction.
As for other future work, those will commence only after whyang fixes the stuffs you outlined, so that's not an issue either.

Do we have a consensus that we want to unblock spawn point implementation and leave the other stuff to whyang, or do you want to have a meeting to discuss this?

**Review by ktpss95112 — APPROVED** [review]

**Comment by ktpss95112** [comment]
OK, let's merge this PR.

**Pull request merged**
